confluent_kafka/schema_registry/schema_registry_client.py

```diff
@@ -154,30 +279,27 @@ def register_schema(self, subject_name, schema):
             https://docs.confluent.io/current/schema-registry/develop/api.html#post--subjects-(string-%20subject)-versions
 
         """  # noqa: E501
-        version = schema.subjects.get(subject_name, None)
-        if version:
-            return version
+        schema_id = self.cache.get_schema_id(schema)
+        if schema_id is not None:
+            return schema_id
 
-        self._rest_client.POST(
-            url_formatter(VERSIONS, subject_name),
-            body=schema, encoder=_SchemaJSONEncoder)
+        response = self._rest_client.POST(
+            'subjects/{}/versions'.format(_urlencode(subject_name)),
+            body=schema)
 
-        # get registration information
-        return self._rest_client.POST(
-            url_formatter(SUBJECT, subject_name),
-            body=schema, encoder=_SchemaJSONEncoder, decoder=_VersionJSONDecoder,
-            schema=schema)
+        schema_id = response['id']
+        self.cache.add_schema(schema_id, schema)
+
+        return schema_id
 
     def get_schema(self, schema_id):
```

Comment:
you will probably want a format parameter, which can be None or 'serialized', otherwise there's no way to get the protobuf base64 encoded schema back. that said i think this might only be necessary for tests.

to implement that, you need something like this guff:

https://github.com/confluentinc/confluent-kafka-dotnet/blob/1.4.x/src/Confluent.SchemaRegistry/CachedSchemaRegistryClient.cs#L319

note that this will then screw up your caching, see the .net for ref impl.